# Actividad - Proyecto práctico


> La actividad se desarrollará en grupos pre-definidos de 2-3 alumnos. Se debe indicar los nombres en orden alfabético (de apellidos). Recordad que esta actividad se corresponde con un 30% de la nota final de la asignatura. Se debe entregar entregar el trabajo en la presente notebook.
*   Alumno 1: Efrain Casas
*   Alumno 2: Manuel Restrepo
*   Alumno 3: Ivan Callejas Sandoval






In [ ]:
print(1)

---
## **PARTE 1** - Instalación y requisitos previos

> Las prácticas han sido preparadas para poder realizarse en el entorno de trabajo de Google Colab. Sin embargo, esta plataforma presenta ciertas incompatibilidades a la hora de visualizar la renderización en gym. Por ello, para obtener estas visualizaciones, se deberá trasladar el entorno de trabajo a local. Por ello, el presente dosier presenta instrucciones para poder trabajar en ambos entornos. Siga los siguientes pasos para un correcto funcionamiento:
1.   **LOCAL:** Preparar el enviroment, siguiendo las intrucciones detalladas en la sección *1.1.Preparar enviroment*.
2.  **AMBOS:** Modificar las variables "mount" y "drive_mount" a la carpeta de trabajo en drive en el caso de estar en Colab, y ejecturar la celda *1.2.Localizar entorno de trabajo*.
3. **COLAB:** se deberá ejecutar las celdas correspondientes al montaje de la carpeta de trabajo en Drive. Esta corresponde a la sección *1.3.Montar carpeta de datos local*.
4.  **AMBOS:** Instalar las librerías necesarias, siguiendo la sección *1.4.Instalar librerías necesarias*.


---
### 1.1. Preparar enviroment (solo local)



> Para preparar el entorno de trabajo en local, se han seguido los siguientes pasos:
1. En Windows, puede ser necesario instalar las C++ Build Tools. Para ello, siga los siguientes pasos: https://towardsdatascience.com/how-to-install-openai-gym-in-a-windows-environment-338969e24d30.
2. Instalar Anaconda
3. Siguiendo el código que se presenta comentado en la próxima celda: Crear un enviroment, cambiar la ruta de trabajo, e instalar librerías básicas.


```
conda create --name miar_rl python=3.8
conda activate miar_rl
cd "PATH_TO_FOLDER"
conda install git
pip install jupyter
```


4. Abrir la notebook con *jupyter-notebook*.



```
jupyter-notebook
```


---
### 1.2. Localizar entorno de trabajo: Google colab o local

In [3]:
# ATENCIÓN!! Modificar ruta relativa a la práctica si es distinta (drive_root)
mount='/content/gdrive'
drive_root = mount + "/My Drive/08_MIAR/actividades/proyecto practico"

try:
  from google.colab import drive
  IN_COLAB=True
except:
  IN_COLAB=False

In [4]:
print(IN_COLAB)

False


---
### 1.3. Montar carpeta de datos local (solo Colab)

In [8]:
# Switch to the directory on the Google Drive that you want to use
import os
if IN_COLAB:
  print("We're running Colab")

  if IN_COLAB:
    # Mount the Google Drive at mount
    print("Colab: mounting Google drive on ", mount)

    drive.mount(mount)

    # Create drive_root if it doesn't exist
    create_drive_root = True
    if create_drive_root:
      print("\nColab: making sure ", drive_root, " exists.")
      os.makedirs(drive_root, exist_ok=True)
    
    # Change to the directory
    print("\nColab: Changing directory to ", drive_root)
    %cd $drive_root
# Verify we're in the correct working directory
%pwd
print("Archivos en el directorio: ")
print(os.listdir())

Archivos en el directorio: 
['.ipynb_checkpoints', '08MIAR_Aprendizaje_por_refuerzo_Proyecto_práctico.ipynb', 'Untitled.ipynb']


---
### 1.4. Instalar librerías necesarias

In [6]:
if IN_COLAB:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install keras-rl2==1.0.5
  %pip install tensorflow==2.8
else:
  %pip install gym==0.17.3
  %pip install git+https://github.com/Kojoley/atari-py.git
  %pip install pyglet==1.5.0
  %pip install h5py==3.1.0
  %pip install Pillow==9.5.0
  %pip install keras-rl2==1.0.5
  %pip install Keras==2.2.4
  %pip install tensorflow==2.5.3
  %pip install torch==2.0.1
  %pip install agents==1.4.0

Note: you may need to restart the kernel to use updated packages.


  Cloning https://github.com/Kojoley/atari-py.git to c:\users\acasa\appdata\local\temp\pip-req-build-sparmquk
  Resolved https://github.com/Kojoley/atari-py.git to commit 86a1e05c0a95e9e6233c3a413521fdb34ca8a089
  Running setup.py clean for atari-py
Failed to build atari-py
  Attempting uninstall: atari-py
    Found existing installation: atari-py 0.2.9
    Uninstalling atari-py-0.2.9:
      Successfully uninstalled atari-py-0.2.9
    Running setup.py install for atari-py: started
    Running setup.py install for atari-py: finished with status 'error'
  Rolling back uninstall of atari-py
  Moving to c:\users\acasa\anaconda3\lib\site-packages\atari_py
   from C:\Users\acasa\anaconda3\Lib\site-packages\~tari_py
  Moving to c:\users\acasa\anaconda3\lib\site-packages\atari_py-0.2.9-py3.9-win-amd64.egg-info
   from C:\Users\acasa\anaconda3\Lib\site-packages\~tari_py-0.2.9-py3.9-win-amd64.egg-info
Note: you may need to restart the kernel to use updated packages.


  Running command git clone -q https://github.com/Kojoley/atari-py.git 'C:\Users\acasa\AppData\Local\Temp\pip-req-build-sparmquk'
  ERROR: Command errored out with exit status 1:
   command: 'C:\Users\acasa\anaconda3\python.exe' -u -c 'import io, os, sys, setuptools, tokenize; sys.argv[0] = '"'"'C:\\Users\\acasa\\AppData\\Local\\Temp\\pip-req-build-sparmquk\\setup.py'"'"'; __file__='"'"'C:\\Users\\acasa\\AppData\\Local\\Temp\\pip-req-build-sparmquk\\setup.py'"'"';f = getattr(tokenize, '"'"'open'"'"', open)(__file__) if os.path.exists(__file__) else io.StringIO('"'"'from setuptools import setup; setup()'"'"');code = f.read().replace('"'"'\r\n'"'"', '"'"'\n'"'"');f.close();exec(compile(code, __file__, '"'"'exec'"'"'))' bdist_wheel -d 'C:\Users\acasa\AppData\Local\Temp\pip-wheel-mzsfi7jx'
       cwd: C:\Users\acasa\AppData\Local\Temp\pip-req-build-sparmquk\
  Complete output (94 lines):
  running bdist_wheel
  running build
  running build_py
  creating build
  creating build\lib.win-amd6

Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.


Note: you may need to restart the kernel to use updated packages.



Note: you may need to restart the kernel to use updated packages.Requirement already satisfied: filelock in c:\users\acasa\anaconda3\lib\site-packages (from torch==2.0.1) (3.3.1)



---
## **PARTE 2**. Enunciado

Consideraciones a tener en cuenta:

- El entorno sobre el que trabajaremos será _SpaceInvaders-v0_ y el algoritmo que usaremos será _DQN_.

- Para nuestro ejercicio, una solución óptima será alcanzada cuando el agente consiga una **media de recompensa por encima de 20 puntos en modo test**. Por ello, esta media de la recompensa se calculará a partir del código de test en la última celda del notebook.

Este proyecto práctico consta de tres partes:

1.   Implementar la red neuronal que se usará en la solución
2.   Implementar las distintas piezas de la solución DQN
3.   Justificar la respuesta en relación a los resultados obtenidos

IMPORTANTE:

* Si no se consigue una puntuación óptima, responder sobre la mejor puntuación obtenida.
* Para entrenamientos largos, recordad que podéis usar checkpoints de vuestros modelos para retomar los entrenamientos. En este caso, recordad cambiar los parámetros adecuadamente (sobre todo los relacionados con el proceso de exploración).

---
## **PARTE 3**. Desarrollo y preguntas

#### Importar librerías

In [6]:
from __future__ import division

from PIL import Image
import numpy as np
import gym

from tensorflow.keras.models import Sequential
from tensorflow.keras.callbacks import Callback
from tensorflow.keras.layers import Dense, Activation, Flatten, Convolution2D, Permute
from tensorflow.keras.optimizers import Adam
import tensorflow.keras.backend as K

from rl.agents.dqn import DQNAgent
from rl.policy import LinearAnnealedPolicy, BoltzmannQPolicy, EpsGreedyQPolicy
from rl.memory import SequentialMemory
from rl.core import Processor
from rl.callbacks import FileLogger, ModelIntervalCheckpoint

#### Configuración base

In [7]:
# Parametros basicos
INPUT_SHAPE = (84, 84)
WINDOW_LENGTH = 4

# Entorno GYM con Videojuego: SpaceInvaders-V0
env_name = 'SpaceInvaders-v0'
env = gym.make(env_name)

# Aleatoriedad Reproducible
np.random.seed(123)
env.seed(123)

# Informacion sobre el entorno
nb_actions = env.action_space.n
print("El tamaño de nuestro frame es: ", env.observation_space)
print("El tamaño de la acción es: ", env.action_space.n)
print("Las acciones disponibles son:")
for action in range(env.action_space.n):
    print("- Acción", action)

In [8]:
# Clase para preprocesar: observaciones, lotes de estados y recompensas

class AtariProcessor(Processor):
    def process_observation(self, observation):
        assert observation.ndim == 3  # (height, width, channel)
        img = Image.fromarray(observation)
        img = img.resize(INPUT_SHAPE).convert('L')
        processed_observation = np.array(img)
        assert processed_observation.shape == INPUT_SHAPE
        return processed_observation.astype('uint8')

    def process_state_batch(self, batch):
        processed_batch = batch.astype('float32') / 255.
        return processed_batch

    def process_reward(self, reward):
        return np.clip(reward, -1., 1.)

In [9]:
# Modificación del RewardThresholdCallback para parar el entrenamiento del agente cuando se logré una recompensa establecida
class RewardThresholdCallback(Callback):
    def __init__(self, reward_threshold):
        super(RewardThresholdCallback, self).__init__()
        self.reward_threshold = reward_threshold

    def on_episode_end(self, episode, logs={}):
        if np.mean(logs['episode_reward']) >= self.reward_threshold:
            self.model.stop_training = True
            print(f"\nTraining stopped. Reached reward threshold of {self.reward_threshold}.")

1. Implementación de la red neuronal

In [10]:
# Next, we build our model. We use the same model that was described by Mnih et al. (2015).
input_shape = (WINDOW_LENGTH,) + INPUT_SHAPE
model = Sequential()
print(K.image_data_format())
if K.image_data_format() == 'channels_last':
    # (width, height, channels)
    model.add(Permute((2, 3, 1), input_shape=input_shape))
elif K.image_data_format() == 'channels_first':
    # (channels, width, height)
    model.add(Permute((1, 2, 3), input_shape=input_shape))
else:
    raise RuntimeError('Unknown image_dim_ordering.')

model.add(Convolution2D(32, (8, 8), strides=(4, 4)))
model.add(Activation('relu'))
model.add(Convolution2D(64, (4, 4), strides=(2, 2)))
model.add(Activation('relu'))
model.add(Convolution2D(64, (3, 3), strides=(2, 2)))
model.add(Activation('relu'))
model.add(Flatten())
model.add(Dense(512))
model.add(Activation('relu'))
model.add(Dense(nb_actions))
model.add(Activation('linear'))
print(model.summary())

channels_last
Model: "sequential"
_________________________________________________________________
Layer (type)                 Output Shape              Param #   
permute (Permute)            (None, 84, 84, 4)         0         
_________________________________________________________________
conv2d (Conv2D)              (None, 20, 20, 32)        8224      
_________________________________________________________________
activation (Activation)      (None, 20, 20, 32)        0         
_________________________________________________________________
conv2d_1 (Conv2D)            (None, 9, 9, 64)          32832     
_________________________________________________________________
activation_1 (Activation)    (None, 9, 9, 64)          0         
_________________________________________________________________
conv2d_2 (Conv2D)            (None, 4, 4, 64)          36928     
_________________________________________________________________
activation_2 (Activation)    (None, 4, 4, 

2. Implementación de la solución DQN

In [18]:
# Hiperparametros

# Hiperparametros del modelo
tasa_aprendizaje = 0.00025         # Alfa (tasa de aprendizaje)

# Parámetros de exploración para la estrategia epsilon-greedy
explorar_inicio = 1.0       # Probabilidad de exploración al inicio
explorar_fin = 0.1          # Probabilidad mínima de exploración
explorar_test = 0.05        # Probabilidad de exploración durante las pruebas
total_pasos = 1000000          # Cantidad total de pasos

# Hiperparámetros de Q-learning
gamma = 0.99                # Tasa de descuento
tasa_actualizacion = 10000  # Frecuencia con la que se actualiza el modelo objetivo
intervalo_entrenamiento = 4 # Intervalo de entrenamiento en el que se actualiza la red neuronal del agente

# Hiperparametros de memoria
longitud_preentrenamiento = 50000  # Número de experiencias almacenadas en la memoria al inicializarse por primera vez
tam_memoria = 1000000           # Número de experiencias que puede almacenar la memoria

# Visualizar el entorno
# Si se ejecuta en Colab, se deja en False ya que aún no hemos encontrado la manera de ver la dinamización del juego que no sea en local
renderizar_episodio = False


# HYPERPARAMETERS CALLBACKS
reward_threshold = 20         #Detener el entrenamiento cuando se obtenga 20 de recomepensa 

---
## **POLÍTICA DE EXPLORACIÓN**

Las siguientes líneas estan relacionadas con la configuración de una política de exploración epsilon-greedy con annealing lineal.

- <strong>LinearAnnealedPolicy:</strong> Esta política permite ajustar el valor de epsilon a lo largo del tiempo para equilibrar la exploración y la explotación durante el entrenamiento.

- <strong>EpsGreedyQPolicy():</strong> Es una política epsilon-greedy básica que selecciona la acción con la mayor valor Q la mayor parte del tiempo y selecciona una acción aleatoria con una probabilidad epsilon.

- <strong>attr='eps':</strong> Especifica que el atributo a ajustar es eps, el cual representa el valor de epsilon.

- <strong>value_max=explorar_inicio:</strong> Corresponde al valor inicial de epsilon (el mayor valor de exploración).

- <strong>value_min=explorar_fin:</strong> corresponde al valor final de epsilon (el menor valor de exploración).

- <strong>value_test=explorar_test: </strong> Corresponde al valor de epsilon a utilizar durante las pruebas o evaluaciones del modelo entrenado.

- <strong>nb_steps=total_pasos: </strong>Especifica el número total de pasos de entrenamiento, lo que determina el intervalo de tiempo en el que se realiza el annealing lineal de epsilon.  Que para nuestro caso lo dejamos en 1.000.000

Como conclusión en la política implementada, el valor de epsilon se va ajustando progresivamente e inicia desde <strong>explorar inicio (1.0)</strong>, hasta <strong>explorar fin (0.1)</strong> y el total de pasos esta definido por la variable <strong>total pasos (1.000.000)</strong>, lo cual nos permite que el modelo de aprendizaje por refuerzo encuentre un equilibrio entre la exploración inicial y la exploración posterior a medida que el entrenamiento avanza.

In [19]:
# Agente DQN (Deep Q-Network)

# Almacenar y recuperar muestras de experiencia
memory = SequentialMemory(limit=tam_memoria, window_length=WINDOW_LENGTH)
# Instancia de la clase AtariProcessor
processor = AtariProcessor()
# Política de exploración
policy = LinearAnnealedPolicy(EpsGreedyQPolicy(), attr='eps',
                              value_max=explorar_inicio, value_min=explorar_fin, value_test=explorar_test,
                              nb_steps=total_pasos)
# Instancia de la clase DQNAgent
dqn = DQNAgent(model=model, nb_actions=nb_actions, policy=policy,
               memory=memory, processor=processor,
               nb_steps_warmup=longitud_preentrenamiento, gamma=gamma,
               target_model_update=tasa_actualizacion,
               train_interval=intervalo_entrenamiento)
# Complicion del Agente
dqn.compile(Adam(learning_rate=tasa_aprendizaje), metrics=['mae'])

In [20]:
# Entrenamiento del Agente

# Guardar los pesos del modelo entrenado
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
checkpoint_weights_filename = 'dqn_' + env_name + '_weights_{step}.h5f'
log_filename = 'dqn_{}_log.json'.format(env_name)
# Callbacks para controlar el entrenamiento del agente
callbacks = [ModelIntervalCheckpoint(checkpoint_weights_filename, interval=250000)]
callbacks += [FileLogger(log_filename, interval=100)]
callbacks += [RewardThresholdCallback(reward_threshold=reward_threshold)]

# Ejecucion del entrenamiento
dqn.fit(env, callbacks=callbacks, nb_steps=1000000, log_interval=10000, visualize=renderizar_episodio)
# Guardar pesos finales del entrenamiento
dqn.save_weights(weights_filename, overwrite=True)

Training for 1000000 steps ...
Interval 1 (0 steps performed)
10000/10000 [==============================] - 35s 3ms/step - reward: 0.0124
14 episodes - episode_reward: 8.571 [4.000, 16.000] - lives: 2.153 - episode_frame_number: 1071.621 - frame_number: 1441048.421

Interval 2 (10000 steps performed)
10000/10000 [==============================] - 35s 4ms/step - reward: 0.0134
16 episodes - episode_reward: 8.312 [3.000, 15.000] - lives: 2.014 - episode_frame_number: 1007.091 - frame_number: 1470937.694

Interval 3 (20000 steps performed)
 5258/10000 [==============>...............] - ETA: 16s - reward: 0.0146
Training stopped. Reached reward threshold of 20.
10000/10000 [==============================] - 35s 4ms/step - reward: 0.0140
13 episodes - episode_reward: 9.923 [5.000, 24.000] - lives: 2.108 - episode_frame_number: 1260.392 - frame_number: 1500794.846

Interval 4 (30000 steps performed)
10000/10000 [==============================] - 38s 4ms/step - reward: 0.0145
14 episodes - e

 2509/10000 [======>.......................] - ETA: 3:54 - reward: 0.0187
Training stopped. Reached reward threshold of 20.
 6273/10000 [=================>............] - ETA: 1:57 - reward: 0.0179
Training stopped. Reached reward threshold of 20.
 7825/10000 [======================>.......] - ETA: 1:08 - reward: 0.0181
Training stopped. Reached reward threshold of 20.
10000/10000 [==============================] - 315s 31ms/step - reward: 0.0178
14 episodes - episode_reward: 12.857 [6.000, 25.000] - loss: 0.014 - mae: 1.615 - mean_q: 1.965 - mean_eps: 0.789 - lives: 2.150 - episode_frame_number: 1182.856 - frame_number: 2129581.999

Interval 25 (240000 steps performed)
 3661/10000 [=========>....................] - ETA: 3:18 - reward: 0.0186
Training stopped. Reached reward threshold of 20.
10000/10000 [==============================] - 325s 32ms/step - reward: 0.0168
15 episodes - episode_reward: 11.400 [6.000, 24.000] - loss: 0.014 - mae: 1.677 - mean_q: 2.038 - mean_eps: 0.780 - li

 1797/10000 [====>.........................] - ETA: 7:20 - reward: 0.0206
Training stopped. Reached reward threshold of 20.
 2793/10000 [=======>......................] - ETA: 6:30 - reward: 0.0211
Training stopped. Reached reward threshold of 20.
 5021/10000 [==============>...............] - ETA: 4:31 - reward: 0.0227
Training stopped. Reached reward threshold of 20.
 5997/10000 [================>.............] - ETA: 3:38 - reward: 0.0228
Training stopped. Reached reward threshold of 20.
 7329/10000 [====================>.........] - ETA: 2:26 - reward: 0.0225
Training stopped. Reached reward threshold of 20.
 9101/10000 [==========================>...] - ETA: 49s - reward: 0.0215
Training stopped. Reached reward threshold of 20.
 9977/10000 [============================>.] - ETA: 1s - reward: 0.0216
Training stopped. Reached reward threshold of 20.
10000/10000 [==============================] - 549s 55ms/step - reward: 0.0216
11 episodes - episode_reward: 19.727 [9.000, 28.000] - l

In [21]:
# Prueba
# Archivo de pesos obtenidos en el entrenamiento
weights_filename = 'dqn_{}_weights.h5f'.format(env_name)
# Cargar los pesos
dqn.load_weights(weights_filename)
# Generar Test
dqn.test(env, nb_episodes=10, visualize=False)

Testing for 10 episodes ...
Episode 1: reward: 28.000, steps: 1366
Episode 2: reward: 22.000, steps: 976
Episode 3: reward: 10.000, steps: 623
Episode 4: reward: 8.000, steps: 610
Episode 5: reward: 15.000, steps: 969
Episode 6: reward: 9.000, steps: 476
Episode 7: reward: 4.000, steps: 358
Episode 8: reward: 8.000, steps: 513
Episode 9: reward: 18.000, steps: 939
Episode 10: reward: 13.000, steps: 662


In [34]:
# Segundo Test
dqn.test(env, nb_episodes=10, visualize=False)

Testing for 10 episodes ...
Episode 1: reward: 19.000, steps: 882
Episode 2: reward: 18.000, steps: 930
Episode 3: reward: 13.000, steps: 712
Episode 4: reward: 8.000, steps: 480
Episode 5: reward: 13.000, steps: 934
Episode 6: reward: 15.000, steps: 768
Episode 7: reward: 12.000, steps: 641
Episode 8: reward: 9.000, steps: 722
Episode 9: reward: 7.000, steps: 478
Episode 10: reward: 16.000, steps: 919


3. Justificación de los parámetros seleccionados y de los resultados obtenidos

De acuerdo con los resultados obtenidos, se evidencia que el en un primer entrenamiento de 1.750.000 y luego en otro de 1.000.000 con algunos pocos cambios, pasos logró alcanzar una recompensa acumulada mayor de 20 en varios episodios durante el entrenamiento. Esto, permite concluir que el agente fue capaz de aprender y mejorar su desempeño a lo largo del proceso de entrenamiento, basado en la arquitectura de CNN implementada y en la configuración de los hiperparametros. Los resultados obtenidos se pueden atribuir a los siguientes factores:

Estrategias de exploración y explotación:

El uso de una política epsilon-greedy con una tasa de exploración inicial alta y una tasa mínima de exploración permitió al agente explorar el entorno y descubrir acciones que conducían a mayores recompensas. En la medida que la tasa de exploración disminuía permitió al agente aprovechar más el conocimiento adquirido y tomar decisiones óptimas (explotación).

Capacidad de generalización:

El modelo de red neuronal utilizado basado en el estado de arte es muy utilizado en varios trabajos y proyectos en los cuales se evidencia buenos resultados (ver referencias). Esta arquitectura permitió al agente generalizar su conocimiento y aplicarlo a nuevas instancias del entorno, lo que condujo a un mejor rendimiento.

Aprendizaje de políticas óptimas:

El algoritmo DQN utilizado es capaz de aprender las políticas óptimas basadas en el valor de acción Q, lo cual se evidencia durante el entrenamiento, en el cual se observa que el agente pudo ajustar continuamente sus estimaciones de valor Q a medida que se encontraba con nuevas experiencias, lo que le permitió aprender y mejorar sus decisiones en función de las recompensas esperadas.

En cuanto al uso de memoria de repetición secuencial permitió al agente almacenar y recordar experiencias pasadas, lo que facilitó un entrenamiento más eficiente y un mejor aprendizaje.

En este ultimo factor, es de gran importancia para el proceso de entrenamiento el cual por ser mediante iteraciones y con el objetivo de obtener una recompensa acumulada mayor de 20 tanto en local como en colab se presentaba algunas limitaciones o como sucedió el entrenamiento tardo un tiempo considerable, algo común en los proyectos de referencia. 

Otra conclusión fue que era necesario tener un callback que detuviera el entrenamiento cuando se alcanzaba una recompensa determinada y no se avanzaba en los entrenamientos siguientes. Es por eso que obtuvimos un mejor rendimiento en el entrenamiento de un millón de pasos vs el entrenamiento de 1.7 millones de pasos.


---